lib/Provider/OAuth2.php

```diff
@@ -0,0 +1,32 @@
+<?php
+
+namespace OCA\SocialLogin\Provider;
+
+use Hybridauth\User;
+
+class OAuth2 extends \Hybridauth\Adapter\OAuth2
+{
+    /**
+    * {@inheritdoc}
+    */
+    public function authenticate()
+    {
+        $this->logger->info(sprintf('%s::authenticate()', get_class($this)));
+
+        if ($this->isConnected()) {
+            return true;
+        }
+
+        $code = filter_input(INPUT_GET, 'code');
+
+        if (empty($code)) {
+            $this->authenticateBegin();
+        }
+        return $this->authenticateFinish();
+    }
+
+    public function getUserProfile(){
+      $userProfile = new User\Profile();
+      return $userProfile;
```

Comment:
What about **required** `$userProfile->identifier` ? Other fields like `email` and `displayName` will be great too.

Reply:
These information are contained in a standard field called `token_id`, which is a JWT token, in the response of this call :
https://github.com/zorn-v/nextcloud-social-login/blob/master/3rdparty/hybridauth/hybridauth/src/Adapter/OAuth2.php#L401

But HybridAuth does not return it. The required `identifier` is there too. I don't think editing the 3rd-party library is a good idea. Do you have an idea ?
We can't use a separate API call, becase this must be a standard OAuth implementation and API calls are specific to the target (github, stackoverflow, etc).

Reply:
I dont think that `token_id` is what we want. `identifier` must identify user. So one user always have the same identifier.

Reply:
this token contains and identifier : https://developers.google.com/identity/protocols/OpenIDConnect#obtainuserinfo

Reply:
There is no "standard" field `token_id` https://tools.ietf.org/html/rfc6749#section-4.1.4

Reply:
Indeed, that's from [OpenID Connect](http://openid.net/specs/openid-connect-core-1_0.html#IDToken), not from OAuth2. I can change the denomination to "OpenID connect", that's still an improvement as HybridAuth does not support it... But I still have the problem of retrieving the JWT token that HybridAuth does not pass.

Reply:
Maybe name it `openid_connect`, not `custom_oauth` then ? To not be confused.

http://openid.net/specs/openid-connect-core-1_0.html#Overview
4. The RP can send a request with the Access Token to the UserInfo Endpoint. 
Maybe make UserInfo Endpoint configurable too ?

Reply:
Yes, I will change the name, it will be better.

The UserInfo endpoint is _recommended*, not _mandatory_, and it's omitted by a lot of openid connect client implementations that consider the token safer.

I think I have no choice but opening a PR on HybridAuth :(

Reply:
You can override some method(s) if you want. Something like
```php
protected function validateAccessTokenExchange($response)
{
    $collection = parent::validateAccessTokenExchange($response);
    if ($collection->exists('token_id')) {
        $tokenId = $collection->get('token_id');
        $data = //process token id;
        $this->storeData('user_data', $data);
    }
}

public function getUserProfile()
{
//...
    $data = getStoredData('user_data');
//...
}
```